File: python/dsa/graphs/bfs.py
# Graphs — Breadth-First Search

"""BFS for unweighted shortest paths and level-order distances.

Time: O(V+E); Space: O(V).
"""

from dsa.graphs.graph import Graph
from dsa.linear.queue import Queue


def distances(graph: Graph, source: int) -> list[int]:
    """Returns distances from source to all vertices (-1 if unreachable).

    Time: O(V+E); Space: O(V).
    """
    n = graph.vertices()
    dist = [-1] * n
    visited = [False] * n
    q = Queue()
    dist[source] = 0
    visited[source] = True
    q.enqueue(source)
    while not q.is_empty():
        u = q.dequeue()
        for edge in graph.neighbors(u):
            v = edge.to
            if not visited[v]:
                visited[v] = True
                dist[v] = dist[u] + 1
                q.enqueue(v)
    return dist

File: java/src/main/java/com/dsa/examples/graphs/BFS.java
// Graphs — Breadth-First Search

package com.dsa.examples.graphs;

import com.dsa.examples.linear.Queue;

import java.util.Arrays;

/**
 * Breadth-first traversal and unweighted shortest paths.
 * Technique: BFS
 * Invariant: Queue processes nodes in nondecreasing distance from source.
 * distances/traverse: Time O(V+E); Space O(V).
 */
public final class BFS {

    private BFS() {
    }

    /**
     * Returns distances from {@code source} to all vertices (-1 if unreachable).
     * Time: O(V+E); Space: O(V).
     */
    public static int[] distances(Graph graph, int source) {
        int n = graph.vertices();
        int[] dist = new int[n];
        Arrays.fill(dist, -1);
        boolean[] visited = new boolean[n];
        Queue<Integer> q = new Queue<>();
        dist[source] = 0;
        visited[source] = true;
        q.enqueue(source);
        while (!q.isEmpty()) {
            int u = q.dequeue();
            for (Graph.Edge e : graph.neighbors(u)) {
                int v = e.to();
                if (!visited[v]) {
                    visited[v] = true;
                    dist[v] = dist[u] + 1;
                    q.enqueue(v);
                }
            }
        }
        return dist;
    }
}
